tools/explicit_ci_jobs.py

```diff
@@ -0,0 +1,134 @@
+import yaml
+import textwrap
+import os
+import subprocess
+import argparse
+
+from typing import Dict, List, Any
+
+REPO_ROOT = os.path.dirname(os.path.dirname(__name__))
+CONFIG_YML = os.path.join(REPO_ROOT, ".circleci", "config.yml")
+WORKFLOWS_DIR = os.path.join(REPO_ROOT, ".github", "workflows")
+
+WORKFLOWS_TO_CHECK = [
+    "binary_builds",
+    "build",
+    "master_build",
+    # These are formatted slightly differently, skip them
+    # "scheduled-ci",
+    # "debuggable-scheduled-ci",
+    # "slow-gradcheck-scheduled-ci",
+    # "ecr_gc",
+    # "promote",
+]
+
+
+def add_job(
+    workflows: Dict[str, Any],
+    workflow_name: str,
+    type: str,
+    job: Dict[str, Any],
+    past_jobs: Dict[str, Any],
+) -> None:
+    """
+    Add job 'job' under 'type' and 'workflow_name' to 'workflow' in place. Also
+    add any dependencies (they must already be in 'past_jobs')
+    """
+    if workflow_name not in workflows:
+        workflows[workflow_name] = {"when": "always", "jobs": []}
+
+    requires = job.get("requires", None)
+    if requires is not None:
+        for reuqirement in requires:
+            dependency = past_jobs[reuqirement]
+            add_job(workflows, dependency["workflow_name"], dependency["type"], dependency["job"], past_jobs)
+
+    workflows[workflow_name]["jobs"].append({type: job})
+
+
+def get_filtered_circleci_config(
+    workflows: Dict[str, Any], relevant_jobs: List[str]
+) -> Dict[str, Any]:
+    """
+    Given an existing CircleCI config, remove every job that's not listed in
+    'relevant_jobs'
+    """
+    new_workflows: Dict[str, Any] = {}
+    past_jobs: Dict[str, Any] = {}
+    for workflow_name, workflow in workflows.items():
+        if workflow_name not in WORKFLOWS_TO_CHECK:
+            # Don't care about this workflow, skip it entirely
+            continue
+
+        for job_dict in workflow["jobs"]:
+            for type, job in job_dict.items():
+                if "name" not in job:
+                    # Job doesn't have a name so it can't be handled
+                    print("Skipping", type)
+                else:
+                    if job["name"] in relevant_jobs:
+                        # Found a job that was specified at the CLI, add it to
+                        # the new result
+                        add_job(new_workflows, workflow_name, type, job, past_jobs)
+
+                    # Record the job in case it's needed as a dependency later
+                    past_jobs[job["name"]] = {
+                        "workflow_name": workflow_name,
+                        "type": type,
+                        "job": job,
+                    }
+
+    return new_workflows
+
+
+def commit_ci(files: List[str], message: str):
```

Comment:
My local `mypy` is complaining about this line:
```
$ mypy --config=mypy-strict.ini tools/explicit_ci_jobs.py
tools/explicit_ci_jobs.py:84:1: error: Function is missing a return type annotation  [no-untyped-def]
Found 1 error in 1 file (checked 1 source file)
```
I'm guessing that's just because this PR's title contains `[skip ci]`?